File: sdk/python/bridge_sdk.py
#!/usr/bin/env python3
from __future__ import annotations
import json
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional


@dataclass
class BridgeCliError(RuntimeError):
    code: str
    message: str
    payload: Optional[dict[str, Any]] = None

    def __str__(self) -> str:
        return f"{self.code}: {self.message}"


class BridgeClient:
    def __init__(self, cli_path: str, workspace: str, *, session_id: str = "sdk-session", client_id: str = "sdk-python") -> None:
        self.cli_path = cli_path
        self.workspace = workspace
        self.session_id = session_id
        self.client_id = client_id

    def _run(self, *args: str) -> dict[str, Any]:
        cmd = [self.cli_path, *args, "--workspace", self.workspace, "--session-id", self.session_id, "--client-id", self.client_id, "--json"]
        proc = subprocess.run(cmd, text=True, capture_output=True)
        try:
            payload = json.loads(proc.stdout)
        except json.JSONDecodeError as exc:
            raise RuntimeError(proc.stderr or proc.stdout or str(exc)) from exc
        if proc.returncode != 0 or not payload.get("ok", False):
            err = payload.get("error", {})
            raise BridgeCliError(err.get("code", "CLI_ERROR"), err.get("message", "request failed"), payload)
        return payload["result"]

    @staticmethod
    def _write_temp(text: str, suffix: str = ".txt") -> str:
        import tempfile
        fd, path = tempfile.mkstemp(prefix="bridge_sdk_", suffix=suffix)
        Path(path).write_text(text, encoding="utf-8")
        return path

    def session_begin(self) -> dict[str, Any]:
        return self._run("session-begin")

    def session_preview(self) -> dict[str, Any]:
        return self._run("session-preview")

    def session_commit(self) -> dict[str, Any]:
        return self._run("session-commit")

    def session_inspect(self) -> dict[str, Any]:
        return self._run("session-inspect")

    def session_drop_change(self, change_id: str) -> dict[str, Any]:
        return self._run("session-drop-change", "--change-id", change_id)

    def session_drop_path(self, path: str) -> dict[str, Any]:
        return self._run("session-drop-path", "--path", path)

    def markdown_upsert_section(self, path: str, heading: str, body: str, *, heading_level: int = 2) -> dict[str, Any]:
        temp = self._write_temp(body, ".md")
        return self._run("markdown-upsert-section", "--path", path, "--heading", heading, "--heading-level", str(heading_level), "--content-file", temp)

    def json_upsert_key(self, path: str, key_path: str, value_json: str) -> dict[str, Any]:
        temp = self._write_temp(value_json, ".json")
        return self._run("json-upsert-key", "--path", path, "--key-path", key_path, "--content-file", temp)

    def yaml_append_item(self, path: str, key_path: str, value_text: str) -> dict[str, Any]:
        temp = self._write_temp(value_text, ".yaml")
        return self._run("yaml-append-item", "--path", path, "--key-path", key_path, "--content-file", temp)

    def html_set_attribute(self, path: str, selector: str, attribute_name: str, attribute_value: str) -> dict[str, Any]:
        return self._run("html-set-attribute", "--path", path, "--query", selector, "--attr-name", attribute_name, "--attr-value", attribute_value)

File: sdk/typescript/bridge_sdk.mjs
import { execFileSync } from 'node:child_process';
import { writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

export class BridgeCliError extends Error {
  constructor(code, message, payload = null) {
    super(`${code}: ${message}`);
    this.code = code;
    this.payload = payload;
  }
}

export class BridgeClient {
  constructor(cliPath, workspace, { sessionId = 'sdk-session', clientId = 'sdk-ts' } = {}) {
    this.cliPath = cliPath;
    this.workspace = workspace;
    this.sessionId = sessionId;
    this.clientId = clientId;
  }

  _run(...args) {
    const output = execFileSync(this.cliPath, [...args, '--workspace', this.workspace, '--session-id', this.sessionId, '--client-id', this.clientId, '--json'], { encoding: 'utf8' });
    const payload = JSON.parse(output);
    if (!payload.ok) {
      const err = payload.error || {};
      throw new BridgeCliError(err.code || 'CLI_ERROR', err.message || 'request failed', payload);
    }
    return payload.result;
  }

  _writeTemp(text, suffix = '.txt') {
    const path = join(tmpdir(), `bridge_sdk_${Date.now()}_${Math.random().toString(16).slice(2)}${suffix}`);
    writeFileSync(path, text, 'utf8');
    return path;
  }

  sessionBegin() { return this._run('session-begin'); }
  sessionPreview() { return this._run('session-preview'); }
  sessionCommit() { return this._run('session-commit'); }
  sessionInspect() { return this._run('session-inspect'); }
  sessionDropChange(changeId) { return this._run('session-drop-change', '--change-id', changeId); }
  sessionDropPath(path) { return this._run('session-drop-path', '--path', path); }

  markdownUpsertSection(path, heading, body, headingLevel = 2) {
    return this._run('markdown-upsert-section', '--path', path, '--heading', heading, '--heading-level', String(headingLevel), '--content-file', this._writeTemp(body, '.md'));
  }

  jsonUpsertKey(path, keyPath, valueJson) {
    return this._run('json-upsert-key', '--path', path, '--key-path', keyPath, '--content-file', this._writeTemp(valueJson, '.json'));
  }

  yamlAppendItem(path, keyPath, valueText) {
    return this._run('yaml-append-item', '--path', path, '--key-path', keyPath, '--content-file', this._writeTemp(valueText, '.yaml'));
  }

  htmlSetAttribute(path, selector, attributeName, attributeValue) {
    return this._run('html-set-attribute', '--path', path, '--query', selector, '--attr-name', attributeName, '--attr-value', attributeValue);
  }
}
